plugins/qtosgrave/qtosgviewer.cpp

```diff
@@ -2112,21 +2127,31 @@ void QtOSGViewer::_UpdateEnvironment()
     }
 }
 
-void QtOSGViewer::_PostToGUIThread(const boost::function<void()>& fn, bool block)
+void QtOSGViewer::_PostToGUIThread(const boost::function<void()>& fn, ViewerCommandPriority priority, bool block)
 {
     if( _nQuitMainLoop != -1 ) {
         // viewer quit, so anything posted won't get processed
         return;
     }
 
     boost::mutex::scoped_lock lockmsg(_mutexGUIFunctions);
-    if( _listGUIFunctions.size() > 1000 ) {
-        // can happen if system is especially slow
-        //RAVELOG_WARN_FORMAT("too many gui post commands, ignoring %d", _listGUIFunctions.size());
+    GUIThreadFunctionPtr pfn(new GUIThreadFunction(fn, block));
+    // Block non-essential fucntions if viewer is not processing any messages
+    if (!_bUpdateEnvironment && priority < ViewerCommandPriority::HIGH) {
+        RAVELOG_WARN("New GUI thread function of priority %d dropped due to _bUpdateEnvironment being set to false", priority);
         return;
     }
-    GUIThreadFunctionPtr pfn(new GUIThreadFunction(fn, block));
-    _listGUIFunctions.push_back(pfn);
+    // GUI thread function queue size limit
+    if (_mapGUIFunctionLists[priority].size() >= _mapGUIFunctionListLimits[priority]) {
+        if(_mapGUIFunctionLists[priority].size() > 0) {
```

Comment:
Is this check necessary? you are inside `_mapGUIFunctionLists[priority].size() >= _mapGUIFunctionListLimits[priority]`